### PR TITLE
8306121: Scene not rendered initially when changing scenes after fix for JDK-8296621

### PR DESCRIPTION
**Pull request by tsayao**
I replicated the fix on #1054 on Linux.

Also fixes [Monocle: Stage no longer gets focus after fix for JDK-8296621](https://bugs.openjdk.org/browse/JDK-8304734)

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed (2 reviews required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer), 1 [Author](https://openjdk.org/bylaws#author))

### Issue
 * [JDK-8306121](https://bugs.openjdk.org/browse/JDK-8306121): Scene not rendered initially when changing scenes after fix for JDK-8296621 (**Bug** - P2)

### Reviewers
 * [Jose Pereda](https://openjdk.org/census#jpereda) (@jperedadnr - **Reviewer**)
 * [Kevin Rushforth](https://openjdk.org/census#kcr) (@kevinrushforth - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jfx.git pull/1110/head:pull/1110` \
`$ git checkout pull/1110`

Update a local copy of the PR: \
`$ git checkout pull/1110` \
`$ git pull https://git.openjdk.org/jfx.git pull/1110/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 1110`

View PR using the GUI difftool: \
`$ git pr show -t 1110`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jfx/pull/1110.diff">https://git.openjdk.org/jfx/pull/1110.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jfx/pull/1110#issuecomment-1518647013)

### CODE REVIEW

**Comment by kevinrushforth** [comment]
I'll look at this early next week. My initial question for you is: Why wasn't the restored call to `requestFocus()` put back in the same place? It used to be in `setScene` and now is in `setVisible`. Was that change necessary?

I'd also like @jperedadnr to review this, since he was looking at [JDK-8304476](https://bugs.openjdk.org/browse/JDK-8304476), which was also caused by the fix for JDK-8296621. I would expect this new fix to also fix JDK-8304476.

/reviewers 2

**Comment by tsayao** [comment]
@kevinrushforth I think it makes sense to call on `setVisible` because it would translate from `stage.show()` and keep the behavior used to have on `setScene`. But on `setScene` it would:

1. In the most common scenario where a `Stage` will only have one `Scene` during it's lifetime, it would call `requestFocus` before the window exists (mapped on the screen). I think it makes more sense to map the window on the screen and then focus. 

2. In the scenario where multiple scenes exists during the lifetime of the `Stage`, it would keep bringing the window up, restoring [JDK-8296621](https://bugs.openjdk.org/browse/JDK-8296621). It could check if it's the first `Scene`, but there's still (1).

There's a scenario where the user would want to `show` the stage but not raise/focus it (like show it on the taskbar, but not pop to the user). I think that should be possible (at least on platforms with window managers). Speaking of this, I think it will not work (will look into it).

**Comment by tsayao** [comment]
Now I see that the call will actually get delayed.

`javafx.stage.Window` will do the calls when the window will be shown(). On line 1162 `WindowStage.setScene` is called:

```
peer.setScene(SceneHelper.getPeer(getScene()));
```

but this is only called on line 1186:

```
peer.setVisible(true);
```

The configuration of Window properties (iconified, maximized, etc) will start on line 1212:

```
WindowHelper.visibleChanged(Window.this, newVisible);
```

which will fall into `Stage` line 1178 `doVisibleChanged`.

I think this is the right place to call `requestFocus` if the window is not minimized/iconified.

Calling on `setScene`  will cause an initially iconified window to be shown (actually pop on the screen) and then be iconified.

**Comment by jperedadnr** [comment]
@tsayao Unrelated to this PR, but I see that your master is [38 commits ahead](https://github.com/tsayao/jfx/compare/openjdk:jfx:master...master) of openjdk:master. 
This is causing that for every new branch and PR that you create, those commits (the first 30 ones date from 3 years ago!) are always added to it.
Could you reset your local master branch to the upstream version?

File: modules/javafx.graphics/src/main/java/javafx/stage/Stage.java [hunk 1]
```diff
@@ -1194,6 +1194,10 @@ private void doVisibleChanged(boolean value) {
             if (peer != null) {
                 peer.setIcons(platformImages);
             }
+
+            if (!isIconified()) {
+                requestFocus();
```
Comment (jperedadnr) [inline modules/javafx.graphics/src/main/java/javafx/stage/Stage.java:1199]: As for Android, and as discussed earlier in comments, this delays a little bit the focus request from the original implementation before JDK-8296621, but it works fine (the stage gets focused), in line with the earlier proposed fix in MonocleWindow:

```
04-24 14:06:45.970 31436 31461 D GraalCompiled: (Original focus request, before JDK-8296621, in WindowStage::setScene)
04-24 14:06:46.395 31436 31461 D GraalCompiled: (Earlier proposed focus request in MonocleWindow::setVisible)
04-24 14:06:46.395 31436 31461 D GraalCompiled:  (Current focus request in Stage::doVisibleChanged)
```

File: modules/javafx.graphics/src/main/java/com/sun/glass/ui/gtk/GtkWindow.java [hunk 2]
```diff
@@ -48,7 +48,7 @@ public GtkWindow(Window owner, Screen screen, int styleMask) {
 
     // empty - not needed by this implementation
     @Override
-    protected void _updateViewSize(long ptr) {}
+    protected native void _updateViewSize(long ptr);
```
Comment (jperedadnr) [inline modules/javafx.graphics/src/main/java/com/sun/glass/ui/gtk/GtkWindow.java:51]: Is there a system test that can be used to verify these changes on Linux? Since you have mentioned this is the same as #1054, I take that SetSceneScalingTest can be used. 

Also, I'm not sure how you would run WinIconified test attached to the JBS ticket to see that it fails or passes. If that can be converted to system test, could you add it to this PR?

File: modules/javafx.graphics/src/main/java/com/sun/glass/ui/gtk/GtkWindow.java [hunk 3]
```diff
@@ -48,7 +48,7 @@ public GtkWindow(Window owner, Screen screen, int styleMask) {
 
     // empty - not needed by this implementation
     @Override
-    protected void _updateViewSize(long ptr) {}
```
Comment (jperedadnr) [inline modules/javafx.graphics/src/main/java/com/sun/glass/ui/gtk/GtkWindow.java:51]: Fix copyright header of modified files

**Review by jperedadnr — COMMENTED** [review]
Tested successfully on Android and Linux. I've added a few comments.

**Comment by tsayao** [comment]
> @tsayao Unrelated to this PR, but I see that your master is [38 commits ahead](https://github.com/tsayao/jfx/compare/openjdk:jfx:master...master) of openjdk:master. This is causing that for every new branch and PR that you create, those commits (the first 30 ones date from 3 years ago!) are always added to it. Could you reset your local master branch to the upstream version?

I was wondering about that (I was not sure how to do it). Fixed it for the next PR.

File: modules/javafx.graphics/src/main/java/com/sun/glass/ui/gtk/GtkWindow.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (tsayao) [reply modules/javafx.graphics/src/main/java/com/sun/glass/ui/gtk/GtkWindow.java:51]: Removed it.

Comment (tsayao) [reply modules/javafx.graphics/src/main/java/com/sun/glass/ui/gtk/GtkWindow.java:51]: Fixed.

File: modules/javafx.graphics/src/main/java/com/sun/glass/ui/gtk/GtkWindow.java (hunk 2 above)
Comment (kevinrushforth) [reply modules/javafx.graphics/src/main/java/com/sun/glass/ui/gtk/GtkWindow.java:50]: I noticed this earlier, and filed [JDK-8298500](https://bugs.openjdk.org/browse/JDK-8298500) to improve our test coverage in this area.

It would be good to make sure that we have a test that does this for at least an inconified window to help validate this PR.

Comment (tsayao) [reply modules/javafx.graphics/src/main/java/com/sun/glass/ui/gtk/GtkWindow.java:50]: I could do a Iconified test, but I think it would not test the case that happened before, the window pops (because of `requestFocus`) and then gets iconified (the user sees the window showing then iconifying).

Comment (kevinrushforth) [reply modules/javafx.graphics/src/main/java/com/sun/glass/ui/gtk/GtkWindow.java:50]: That seems good enough to verify that the window doesn't stay showing.

Comment (tsayao) [reply modules/javafx.graphics/src/main/java/com/sun/glass/ui/gtk/GtkWindow.java:50]: @kevinrushforth I was checking and `SceneChangeShouldNotFocusStageTest` added on [JDK-8296621](https://bugs.openjdk.org/browse/JDK-8296621) does exactly that.

But I noticed the same as you described on [JDK-8298500](https://bugs.openjdk.org/browse/JDK-8298500).

@jperedadnr The `WinIconified.java` attached is to show that a window should stay iconified and that it is initially iconified (does not pop on the screen and then gets iconified).

**Review by lukostyra — COMMENTED** [review]
I verified the changes, in general the `testSecondSetScene` test is now passing on all platforms. However, on Windows and macOS I observed two tests from controls test suite failing:
* MouseEventFirerTest. testAnchorRight[0]
* ComboBoxTest. test_rt28245

When I change branches to before all these changes - 48b5fea374 - or even before the most recent master that's merged to these tests - 559ad1139b - these tests pass. Seems like these two failures were introduced by this PR.

File: modules/javafx.graphics/src/main/java/com/sun/glass/ui/gtk/GtkWindow.java (hunk 2 above)
Comment (tsayao) [reply modules/javafx.graphics/src/main/java/com/sun/glass/ui/gtk/GtkWindow.java:50]: `SceneChangeShouldNotFocusStageTest` covers the test that the window should start iconified. It does not cover the test where the window pops and then iconifies (causing a visual "weirdness"). I don't think the later is possible on a systemTest.

Comment (jperedadnr) [reply modules/javafx.graphics/src/main/java/com/sun/glass/ui/gtk/GtkWindow.java:50]: If not as system test, then could you add `WinIconified` as manual test, with some javadoc as quick instructions on how to check that it works as expected?

Comment (tsayao) [reply modules/javafx.graphics/src/main/java/com/sun/glass/ui/gtk/GtkWindow.java:50]: @jperedadnr added.

**Comment by tsayao** [comment]
I don't know why `ComboBoxTest.test_rt28245` was working before, but it had just an ENTER to select an element. I fixed it by using the combination of keys to select the first element.

**Comment by tsayao** [comment]
The `MouseEventFirerTest. testAnchorRight[0]` is failing because the `requestFocus` call:

```
  if (!isIconified()) {
      peer.requestFocus();
  }
```

Weird..

**Comment by tsayao** [comment]
I think the `testAnchorRight` is failing because the change exposed a bug somewhere else. The problem seems to be on the `topLeft.getBoundsInParent().getMinX()` function that returned `556`  before and is returning `554.5999755859375` now. So `requestFocus` causes some recalculation that is probably wrong.

**Comment by tsayao** [comment]
I know what happened, it's the focus shadow of the button `topLeft` - it was the first child of the AnchorPane, so it got the focus and hence the difference in the value. Moving it in the Scene graph fixed it (because it does not get the primary focus).

**Comment by tsayao** [comment]
@lukostyra those tests that were failing are passing now.

File: modules/javafx.graphics/src/main/java/com/sun/glass/ui/gtk/GtkWindow.java (hunk 2 above)
Comment (jperedadnr) [reply modules/javafx.graphics/src/main/java/com/sun/glass/ui/gtk/GtkWindow.java:50]: Thanks, @tsayao 

I'm running these tests now on Linux (Ubuntu 20.04) and MacOS (13.3.1):

- controls:test
- Manual/ StartIconified
- System/Robot test.robot.javafx.stage.IconifyTest
- System/Robot test.robot.javafx.scene.SceneChangeShouldNotFocusStageTest

On Linux, all of them work fine.
On MacOS:
 - Work fine
 - I see the stage, it is not iconified (with or without your patch). Is this a know issue?
 - Works fine
 - Fails: it is not iconified (with or without your patch, also I had another issue with scene being null). Is it a known issue as well?

Comment (tsayao) [reply modules/javafx.graphics/src/main/java/com/sun/glass/ui/gtk/GtkWindow.java:50]: I don't have a mac to test it. Was it working before?

Comment (jperedadnr) [reply modules/javafx.graphics/src/main/java/com/sun/glass/ui/gtk/GtkWindow.java:50]: Hmm, tested 17 to 21-ea, it doesn't work (StartIconified always shows the stage on macOS)

**Comment by kevinrushforth** [comment]
> About the `testAnchorRight` test: I know what happened, it's the focus shadow of the button `topLeft` - it was the first child of the AnchorPane, so it got the focus and hence the difference in the value. Moving it in the Scene graph fixed it (because it does not get the primary focus).

This means that there will be a sometimes-noticeable difference in behavior between JavaFX 19 (before the fix for [JDK-8296621](https://bugs.openjdk.org/browse/JDK-8296621)) and JavaFX 21 (after this fix for [JDK-8306121](https://bugs.openjdk.org/browse/JDK-8306121)). I don't think this is a problem, but I'd like to take a closer look.

**Comment by tsayao** [comment]
For some reason `Platform.startup` is hanging and any test that uses it gets stuck.

**Comment by tsayao** [comment]
The problem happens only when `Scene` is created without size, so it expands to `Window` size. If `Scene` has the size specified, the bug does not happen.

**Comment by tsayao** [comment]
Maybe it's possible to do a Java fix where the new Scene without size inherits the old Scene size? This way I think it possible to remove platform specific code such as the one I did and `_updateViewSize`.

**Comment by kevinrushforth** [comment]
I ran an automated test job using your latest patch and all of the tests now pass.

> Maybe it's possible to do a Java fix where the new Scene without size inherits the old Scene size? This way I think it possible to remove platform specific code such as the one I did and `_updateViewSize`.

Hmm. That might be problematic. In any case, I don't want to go down that route to fix this bug.

I have two questions on the current version of the fix:

1. Can you explain why the two tests had to be changed? Unless the tests in question really were buggy, and just getting luck before, the changes suggest a change in behavior that applications might notice and be bothered by.
2. Are you sure that on Linux, there are no cases where `_updateViewSize` needs to be implemented? Did you consider just leaving it as-is or would that have failed?

**Comment by tsayao** [comment]
> I ran an automated test job using your latest patch and all of the tests now pass.
> 
> > Maybe it's possible to do a Java fix where the new Scene without size inherits the old Scene size? This way I think it possible to remove platform specific code such as the one I did and `_updateViewSize`.
> 
> Hmm. That might be problematic. In any case, I don't want to go down that route to fix this bug.
> 
> I have two questions on the current version of the fix:
> 
>     1. Can you explain why the two tests had to be changed? Unless the tests in question really were buggy, and just getting luck before, the changes suggest a change in behavior that applications might notice and be bothered by.
> 
>     2. Are you sure that on Linux, there are no cases where `_updateViewSize` needs to be implemented? Did you consider just leaving it as-is or would that have failed?

I changed it to use `_updateViewSize` so it's compatible with other platforms.

I'm gathering some information to answer about the test changes.

**Comment by kevinrushforth** [comment]
> I changed it to use `_updateViewSize` so it's compatible with other platforms.

As it turns out, I was looking at your incremental diff and (mistakenly) thought that `_updateViewSize` was in the current mainline code base and that you had removed it, which is why I asked about it. I now see that what happened is that the initial version of your PR had added it, and a later version had reverted that addition. And your most recent version added it back in. I don't mind either way, so if you think it's cleaner to implement `_updateViewSize` on Linux (even though today it's only implemented on Windows), that's OK. If you want to revert this latest addition of that method, that's OK, too.

> I'm gathering some information to answer about the test changes.

Thanks.

**Comment by tsayao** [comment]
It seems Headless monocle was not focusing the window before, so this is why tests were passing before (focus was not being accounted).

I've attached a simple `FocusTest` on JBS that prints to the console when Stage/TextField are focused. It does not print anything before this change.

`java -Dglass.platform=Monocle -Dmonocle.platform=Headless -Dprism.order=sw @build/testrun.args FocusTest.java`

**Comment by tsayao** [comment]
I confirm that even restoring [JDK-8296621](https://bugs.openjdk.org/browse/JDK-8296621), the tests were not getting focus.

I did this by adding to `MouseEventFirerTest` on `setup`:

```
        topLeft.focusedProperty().addListener((observable, oldValue, newValue) -> {
            System.out.println("topLeft focused " + newValue);
        });

        stage.focusedProperty().addListener((observable, oldValue, newValue) -> {
            System.out.println("Stage focused " + newValue);
        });

```

And adding `testLogging.showStandardStreams = true;` to `build.gradle`

With this PR, focus messages are shown, without the PR or restoring  [JDK-8296621](https://bugs.openjdk.org/browse/JDK-8296621) they don't.

Probably something to do with Monocle flow.

**Comment by tsayao** [comment]
@kevinrushforth Got it.

Testing uses `StubToolkit` and not `QuantumToolkit` which before [JDK-8296621](https://bugs.openjdk.org/browse/JDK-8296621) called `requestFocus()` inside `WindowStage` which is part of `QuantumToolkit`. So focusing were never taken into account while testing.

**Review by kevinrushforth — APPROVED** [review]
Thanks for the explanation about why you had to change the tests.

The latest fix looks good. All my testing is green.

Running the new `StartIconified` manual test reveals two problems:
1. On Windows there is a brief flash where the window is visible before it is iconified -- we should file a P4 bug
2. On macOS 13 (not sure about 12) the Stage is not iconified at all -- this is tracked by [JDK-8305675](https://bugs.openjdk.org/browse/JDK-8305675)

File: modules/javafx.graphics/src/main/java/com/sun/glass/ui/gtk/GtkWindow.java [hunk 4]
```diff
@@ -46,8 +46,9 @@ public GtkWindow(Window owner, Screen screen, int styleMask) {
     @Override
     protected native boolean _setView(long ptr, View view);
 
+    // empty - not needed by this implementation
     @Override
-    protected native void _updateViewSize(long ptr);
+    protected void _updateViewSize(long ptr) {}
```
[the remark this answers is not in the document]
Comment (kevinrushforth) [reply modules/javafx.graphics/src/main/java/com/sun/glass/ui/gtk/GtkWindow.java]: The above was an old comment, please ignore it.

**Comment by kevinrushforth** [comment]
@jperedadnr This should now be ready for you to resume your review.

**Comment by kevinrushforth** [comment]
I filed https://bugs.openjdk.org/browse/JDK-8310029 to track the brief flash on Windows platforms.

File: tests/manual/stage/StartIconified.java [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+import javafx.scene.text.Text;
+import javafx.stage.Stage;
+import javafx.scene.Scene;
+import javafx.scene.layout.StackPane;
+
+/***
+ * Stage must initially only show on the OS taskbar, but not on the Screen.
+ * If the stage pops on the Screen and then iconifies, it's wrong.
+ */
+public class StartIconified extends Application {
```
Comment (jperedadnr) [inline tests/manual/stage/StartIconified.java:37]: Should we mention this doesn't work on macOS until https://bugs.openjdk.org/browse/JDK-8305675 gets fixed?

**Review by jperedadnr — COMMENTED** [review]
Looks good
Tested on macOS and Linux (with known issue on macOS), and Android, so indeed https://bugs.openjdk.org/browse/JDK-8304476 can be closed as duplicated.

I haven't tested Monocle/TestFX though.

I have minor comments only.

**Comment by tsayao** [comment]
@jperedadnr Did the changes requested in review.

**Review by jperedadnr — APPROVED** [review]
Looks good to me!

File: modules/javafx.controls/src/test/java/test/com/sun/javafx/scene/control/infrastructure/MouseEventFirerTest.java [hunk 6]
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2020, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2023, Oracle and/or its affiliates. All rights reserved.
```
Comment (kevinrushforth) [inline modules/javafx.controls/src/test/java/test/com/sun/javafx/scene/control/infrastructure/MouseEventFirerTest.java:2]: No, we never remove the start year in a copyright line. this should be: `2020, 2023,`

Alternatively, you can just revert this change, since we will run a script to update them all. Either is fine.

File: modules/javafx.controls/src/test/java/test/javafx/scene/control/ComboBoxTest.java [hunk 7]
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2011, 2022, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2011, 2023, Oracle and/or its affiliates. All rights reserved.
```
Comment (kevinrushforth) [inline modules/javafx.controls/src/test/java/test/javafx/scene/control/ComboBoxTest.java:2]: `2022, 2023,`

**Review by kevinrushforth — CHANGES_REQUESTED** [review]
Copyright year changes are incorrect.

File: modules/javafx.controls/src/test/java/test/javafx/scene/control/ComboBoxTest.java (hunk 7 above)
Comment (jperedadnr) [reply modules/javafx.controls/src/test/java/test/javafx/scene/control/ComboBoxTest.java:2]: this is one is correct, though? `2011, 2023,` should be fine?

**Review by kevinrushforth — APPROVED** [review]

**Review by jperedadnr — APPROVED** [review]

File: modules/javafx.graphics/src/main/java/javafx/stage/Stage.java [hunk 8]
```diff
@@ -1187,6 +1187,10 @@ private void doVisibleChanged(boolean value) {
             peer.setMaximized(isMaximized());
             peer.setTitle(getTitle());
 
+            if (!isIconified()) {
+                peer.requestFocus();
+            }
```
[the remark this answers is not in the document]
Comment (kevinrushforth) [reply modules/javafx.graphics/src/main/java/javafx/stage/Stage.java:1192]: Thanks for confirming. I just now closed [JDK-8304476](https://bugs.openjdk.org/browse/JDK-8304476) as a duplicate of [JDK-8306121](https://bugs.openjdk.org/browse/JDK-8306121).

**Comment by kevinrushforth** [comment]
I _think_ [JDK-8304734](https://bugs.openjdk.org/browse/JDK-8304734) can be closed as a duplicate of this bug, but it would be helpful to confirm that it resolves the problem reported in TestFX/TestFX#749.
